lib/web_ui/lib/src/engine/html/color_filter.dart

```diff
@@ -0,0 +1,286 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.10
+part of engine;
+
+/// A surface that applies an [ColorFilter] to its children.
+class PersistedColorFilter extends PersistedContainerSurface
+    implements ui.ColorFilterEngineLayer {
+  PersistedColorFilter(PersistedColorFilter? oldLayer, this.filter) : super(oldLayer);
+
+  @override
+  html.Element? get childContainer => _childContainer;
+  /// The dedicated child container element that's separate from the
+  /// [rootElement] is used to compensate for the coordinate system shift
+  /// introduced by the [rootElement] translation.
+  html.Element? _childContainer;
+
+  final ui.ColorFilter filter;
+  html.Element? _filterElement;
+  bool containerVisible = true;
+
+  @override
+  void adoptElements(PersistedColorFilter oldSurface) {
+    super.adoptElements(oldSurface);
+    _childContainer = oldSurface._childContainer;
+    oldSurface._childContainer = null;
+  }
+
+  @override
+  void discard() {
+    super.discard();
+    // Do not detach the child container from the root. It is permanently
+    // attached. The elements are reused together and are detached from the DOM
+    // together.
+    _childContainer = null;
+  }
+
+  @override
+  html.Element createElement() {
+    html.Element element = defaultCreateElement('flt-color-filter');
+    _childContainer = html.Element.tag('flt-filter-interior');
+    element.append(_childContainer!);
+    return element;
+  }
+
+  @override
+  void apply() {
+    if (_filterElement != null) {
+      _filterElement?.remove();
+    }
+    final EngineColorFilter? engineValue = filter as EngineColorFilter?;
+    if (engineValue == null) {
+      rootElement!.style.backgroundColor = '';
+      childContainer?.style.visibility = 'visible';
+      return;
+    }
+    if (engineValue._blendMode == null) {
+      rootElement!.style.backgroundColor = colorToCssString(engineValue._color!);
+      childContainer?.style.visibility = 'visible';
+      return;
+    }
+
+    ui.Color filterColor = engineValue._color!;
+    ui.BlendMode? colorFilterBlendMode = engineValue._blendMode;
+    html.CssStyleDeclaration style = rootElement!.style;
+    if (colorFilterBlendMode != null) {
+      switch (colorFilterBlendMode) {
+        case ui.BlendMode.clear:
+        case ui.BlendMode.dstOut:
+        case ui.BlendMode.srcOut:
+          childContainer?.style.visibility = 'hidden';
+          return;
+        case ui.BlendMode.dst:
+        case ui.BlendMode.dstIn:
+          // Noop.
```

Comment:
Is it safe to keep the old value of `childContainer.style.visibility`? Should we set it to an empty string?